controllers/tykoas_controller.go

```diff
@@ -0,0 +1,324 @@
+	if err != nil {
+		return "", err
+	}
+
+	if apiID == "" {
+		apiID = EncodeNS(client.ObjectKeyFromObject(tykOASCrd).String())
+	}
+
+	exists := klient.Universal.TykOAS().Exists(ctx, apiID)
+	if !exists {
+		if err = klient.Universal.TykOAS().Create(ctx, apiID, tykOASDoc); err != nil {
+			r.Log.Error(err, "Failed to create Tyk OAS API", "name", client.ObjectKeyFromObject(tykOASCrd))
+			return "", err
+		}
+	} else {
+		if err = klient.Universal.TykOAS().Update(ctx, apiID, tykOASDoc); err != nil {
+			r.Log.Error(err, "Failed to update Tyk OAS API", "name", client.ObjectKeyFromObject(tykOASCrd))
+			return "", err
+		}
+	}
+
+	return apiID, nil
+}
+
+func (r *TykOASReconciler) updateStatus(ctx context.Context, tykOASCrd *v1alpha1.TykOAS, apiID string) error {
+	var cm v1.ConfigMap
+
+	r.Log.Info("Updating status of Tyk OAS instance")
+
+	cmNS := types.NamespacedName{
+		Name:      tykOASCrd.Spec.TykOAS.ConfigmapRef.Name,
+		Namespace: tykOASCrd.Spec.TykOAS.ConfigmapRef.Namespace,
+	}
+
+	err := r.Client.Get(ctx, cmNS, &cm)
+	if err != nil {
+		r.Log.Error(err, "Failed to fetch config map", "name", cmNS)
+		return err
+	}
+
+	tykOASDoc := cm.Data[tykOASCrd.Spec.TykOAS.ConfigmapRef.KeyName]
+
+	ns := client.ObjectKeyFromObject(tykOASCrd)
+
+	if tykOASCrd.Status.ID == "" {
+		tykOASCrd.Status.ID = apiID
+	}
+
+	state, err := jsonparser.GetBoolean([]byte(tykOASDoc), TykOASExtenstionStr, "info", "state", "active")
+	if err != nil && err != jsonparser.KeyPathNotFoundError {
+		r.Log.Error(err, "Failed to fetch state information from Tyk OAS document", "name", ns.String())
+	} else {
+		tykOASCrd.Status.Enabled = state
+	}
+
+	str, err := jsonparser.GetString([]byte(tykOASDoc), TykOASExtenstionStr, "server", "customDomain")
+	if err != nil && err != jsonparser.KeyPathNotFoundError {
+		r.Log.Error(err, "Failed to fetch domain information from Tyk OAS document", "name", ns.String())
+	} else {
+		tykOASCrd.Status.Domain = str
+	}
+
+	str, err = jsonparser.GetString([]byte(tykOASDoc), TykOASExtenstionStr, "server", "listenPath", "value")
+	if err != nil && err != jsonparser.KeyPathNotFoundError {
+		r.Log.Error(err, "Failed to fetch listen path information from Tyk OAS document", "name", ns.String())
+	} else {
+		tykOASCrd.Status.ListenPath = str
+	}
+
+	str, err = jsonparser.GetString([]byte(tykOASDoc), TykOASExtenstionStr, "upstream", "url")
+	if err != nil && err != jsonparser.KeyPathNotFoundError {
+		r.Log.Error(err, "Failed to fetch upstream url  information from Tyk OAS document", "name", ns.String())
+	} else {
+		tykOASCrd.Status.TargetURL = str
+	}
+
+	return r.Client.Status().Update(ctx, tykOASCrd)
+}
+
+func (r *TykOASReconciler) getAPIID(tykOASCrd *v1alpha1.TykOAS, tykOASDoc string) (string, error) {
+	if tykOASCrd.Status.ID != "" {
+		return tykOASCrd.Status.ID, nil
+	}
+
+	val, err := jsonparser.GetString([]byte(tykOASDoc), TykOASExtenstionStr, "info", "id")
+	if err != nil && err != jsonparser.KeyPathNotFoundError {
```

Comment:
`KeyPathNotFoundError` is ignored because of that `id` may not exists on all OAS docs, right? Maybe we can add a comment line here or method level to describe this case.